backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -365,4 +418,362 @@ public FutureExperimentRunDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
+    if (builder.hasPython()) {
+      PythonEnvironmentBlob.Builder pythonEnvironmentBlobBuilder = builder.getPython().toBuilder();
+
+      // Compare requirementEnvironmentBlobs
+      List<PythonRequirementEnvironmentBlob> requirementEnvironmentBlobs =
+          new ArrayList<>(pythonEnvironmentBlobBuilder.getRequirementsList());
+      requirementEnvironmentBlobs.sort(
+          Comparator.comparing(PythonRequirementEnvironmentBlob::getLibrary));
+      pythonEnvironmentBlobBuilder
+          .clearRequirements()
+          .addAllRequirements(requirementEnvironmentBlobs);
+
+      // Compare
+      List<PythonRequirementEnvironmentBlob> constraintsBlobs =
+          new ArrayList<>(pythonEnvironmentBlobBuilder.getConstraintsList());
+      constraintsBlobs.sort(Comparator.comparing(PythonRequirementEnvironmentBlob::getLibrary));
+      pythonEnvironmentBlobBuilder.clearConstraints().addAllConstraints(constraintsBlobs);
+
+      builder.setPython(pythonEnvironmentBlobBuilder.build());
+    }
+    return builder.build();
+  }
+
+  private InternalFuture<ExperimentRun> checkIfEntityAlreadyExists(
+      ExperimentRun experimentRun, Boolean isInsert) {
+    return jdbi.useHandle(
+            handle -> {
+              try {
+                String queryStr;
+                if (isInsert) {
+                  queryStr =
+                      "SELECT id FROM experiment_run ere WHERE "
+                          + " ere."
+                          + ModelDBConstants.NAME
+                          + " = :experimentRunName "
+                          + " AND ere."
+                          + ModelDBConstants.PROJECT_ID
+                          + " = :projectId "
+                          + " AND ere."
+                          + ModelDBConstants.EXPERIMENT_ID
+                          + " = :experimentId "
+                          + " AND ere."
+                          + ModelDBConstants.DELETED
+                          + " = false ";
+                } else {
+                  queryStr =
+                      "SELECT id FROM experiment_run ere WHERE "
+                          + " ere."
+                          + ModelDBConstants.ID
+                          + " = :experimentRunId "
+                          + " AND ere."
+                          + ModelDBConstants.DELETED
+                          + " = false ";
+                }
+
+                var query = handle.createQuery(queryStr);
+
+                if (isInsert) {
+                  query.bind("experimentRunName", experimentRun.getName());
+                  query.bind("projectId", experimentRun.getProjectId());
+                  query.bind("experimentId", experimentRun.getExperimentId());
+                } else {
+                  query.bind(ModelDBConstants.EXPERIMENT_RUN_ID_STR, experimentRun.getId());
+                }
+
+                long count = query.mapTo(String.class).stream().count();
```

Comment:
```suggestion
                Long count = query.mapTo(Long.class).one();
```
Push the count to the DB. Don't do the count in Java. It's wasteful.

https://github.com/VertaAI/docs/blob/main/engineering-practices/api-implementation/performance.md#simple-db-operations-with-indices

Reply:
updated